src/model_pool.hpp

```diff
@@ -157,203 +147,180 @@ namespace wr
 
 	//! Loads a model without materials
 	template<typename TV, typename TI>
-	Model* ModelPool::Load(MaterialPool* material_pool, TexturePool* texture_pool, std::string_view path, ModelType type)
+	Model* ModelPool::Load(MaterialPool* material_pool, TexturePool* texture_pool, std::string_view path)
 	{
-		IS_PROPER_VERTEX_CLASS(TV)
-
-		Assimp::Importer importer;
-		const aiScene* scene = importer.ReadFile(path.data(),
-			aiProcess_Triangulate |
-			aiProcess_CalcTangentSpace |
-			aiProcess_JoinIdenticalVertices |
-			aiProcess_OptimizeMeshes |
-			aiProcess_ImproveCacheLocality |
-			aiProcess_MakeLeftHanded);
-
-		if (!scene || scene->mFlags & AI_SCENE_FLAGS_INCOMPLETE || !scene->mRootNode)
+		IS_PROPER_VERTEX_CLASS(TV);
+
+		AssimpModelLoader assimp_loader = AssimpModelLoader();
+		ModelLoader* loader;
+
+		if (assimp_loader.SupportsModelFormat(path.substr(path.find_last_of(".") + 1).data()))
 		{
-			LOGW(std::string("Loading model ") +
-				path.data() +
-				std::string(" failed with error ") +
-				importer.GetErrorString());
-			return nullptr;
+			loader = new AssimpModelLoader();
 		}
 
+		ModelData* data = loader->Load(path);
+
 		Model* model = new Model;
 
 		MaterialHandle* default_material = nullptr;
 
 		// TODO: Create default material
 
-		int ret = LoadNodeMeshes<TV, TI>(scene, scene->mRootNode, model, default_material);
+		int ret = LoadNodeMeshes<TV, TI>(data, model, default_material);
 
 		if (ret == 1)
 		{
 			DestroyModel(model);
+
+			delete loader;
 			return nullptr;
 		}
 		model->m_model_name = path.data();
 		model->m_model_pool = this;
 
+		delete loader;
+
 		return model;
 	}
 
 	//! Loads a model with materials
 	template<typename TV, typename TI>
-	Model* ModelPool::LoadWithMaterials(MaterialPool* material_pool, TexturePool* texture_pool, std::string_view path, ModelType type)
+	Model* ModelPool::LoadWithMaterials(MaterialPool* material_pool, TexturePool* texture_pool, std::string_view path)
 	{
-		IS_PROPER_VERTEX_CLASS(TV)
-
-		Assimp::Importer importer;
-		const aiScene* scene = importer.ReadFile(path.data(),
-			aiProcess_Triangulate |
-			aiProcess_CalcTangentSpace |
-			aiProcess_JoinIdenticalVertices |
-			aiProcess_OptimizeMeshes |
-			aiProcess_ImproveCacheLocality |
-			aiProcess_MakeLeftHanded);
-
-		if (!scene || scene->mFlags & AI_SCENE_FLAGS_INCOMPLETE || !scene->mRootNode)
+		IS_PROPER_VERTEX_CLASS(TV);
+
+		AssimpModelLoader assimp_loader();
```

Comment:
Consider static initialization of model loaders.

Reply:
Done. You still have to create the actual class somewhere. I'm currently using demo.cpp for that.